red4ext-sys/src/interop.rs

```diff
@@ -84,7 +84,104 @@ unsafe impl ExternType for TweakDBID {
     type Kind = cxx::kind::Trivial;
 }
 
-#[derive(Debug, Clone)]
+/// see [its C++ representation](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/NativeTypes.hpp#L105)
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(C)]
+pub struct ItemID {
+    id: TweakDBID,
+    seed: Seed,
+    counter: u16,
+    /// also called [`unknown` in CET](https://wiki.redmodding.org/cyber-engine-tweaks/functions/special-types#toitemid)
+    structure: gamedataItemStructure,
+    /// also called [`maybe_type` in CET](https://wiki.redmodding.org/cyber-engine-tweaks/functions/special-types#toitemid)
+    flags: gameEItemIDFlag,
+}
+
+impl ItemID {
+    pub fn new() -> Self {
+        Self {
+            ..Default::default()
+        }
+    }
+    pub fn new_from_tdbid(id: &TweakDBID, seed: impl Into<Seed>) -> Self {
+        Self {
+            id: id.clone(),
+            seed: seed.into(),
+            ..Default::default()
+        }
+    }
+    pub fn new_from_params(
+        id: &TweakDBID,
+        seed: impl Into<Seed>,
+        counter: u16,
+        structure: gamedataItemStructure,
+        flags: gameEItemIDFlag,
+    ) -> Self {
+        Self {
+            id: id.clone(),
+            seed: seed.into(),
+            counter,
+            structure,
+            flags,
+        }
+    }
+}
+
+/// see [gameEItemIDFlag](https://nativedb.red4ext.com/gameEItemIDFlag)
+#[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd, Eq)]
+#[repr(u8)]
+#[allow(non_camel_case_types, dead_code)]
+pub enum gameEItemIDFlag {
+    #[default]
+    /// see [CET initialization](https://github.com/maximegmd/CyberEngineTweaks/blob/v1.24.1/src/scripting/Scripting.cpp#L311)
+    None = 0,
+    Preview = 1,
+}
+
+unsafe impl ExternType for gameEItemIDFlag {
+    type Id = type_id!("RED4ext::gameEItemIDFlag");
+    type Kind = cxx::kind::Trivial;
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd, Eq)]
+#[repr(u8)]
+#[allow(non_camel_case_types, dead_code)]
+pub enum gamedataItemStructure {
```

Comment:
IMO we should use the rust naming convention and capitalize it

Reply:
I think if you put the enum inside the ffi module cxx will generate an assert to make sure the variants are correct:
https://cxx.rs/shared.html#extern-enums
(and you can omit impl ExternType)

Reply:
I've moved 1 enum if you need an example:
https://github.com/jac3km4/red4ext-rs/commit/dc08014be6107aeac6d8d121355989b277784779

Reply:
something that I quite don't get : if I place e.g. type gamedataItemStructure under #[namespace = "RED4ext::Memory"] as you did for EMainReason then the compilation fails from cxx as it seems.

Reply:
ok I've just tested on my local `red4ext-rs-example` and indeed in vanilla an `ItemID` is only created if valid, otherwise both its `id` and `seed` are always **0**.
here's a screenshot, with a simple test func:

```swift
@addMethod(PlayerPuppet)
public func TestCreateItemID() -> Void {
 let name = "Items.FirstAidWhiffV0";
 let tdbid = TDBID.Create(name);
 let vanilla = ItemID.FromTDBID(tdbid);
 let mod = CreateItemID(tdbid);


 LogChannel(n"DEBUG", s"\(ToString(vanilla)) \(TDBID.ToStringDEBUG(ItemID.GetTDBID(vanilla)))");
 LogChannel(n"DEBUG", s"\(ToString(mod)) \(TDBID.ToStringDEBUG(ItemID.GetTDBID(mod)))");
}
```
here's what it yields, with 2 different invalid `String` and 1 valid:
![fix-itemid](https://user-images.githubusercontent.com/21016014/233765238-e49ba7c1-da0c-443e-bee0-06946ed790c9.png)
it shows that in order to allow creating `ItemID` in the first, internally we should have an implementation for `IsValid`:
https://jac3km4.github.io/cyberdoc/#10126
https://jac3km4.github.io/cyberdoc/#15890

also note that currently `call!` is not defined at type definition site, so `red4ext-rs` cannot call `redscript` there.

Reply:
@Roms1383 you have to make sure that the `gamedataItemStructure` type is visible from C++, which means you might need to include an additional file by adding something like
`include!("RED4ext/Scripting/Natives/Generated/game/data/ItemStructure_Record.hpp")`

Reply:
If you want to add `IsValid` to `ItemID` I think it'd be preferable to implement it manually, since that's less error prone.
When I talked about calling `IsValid` using the macro I just meant that a user can pretty easily call the method, so if it's not included in the crate, it's not that big of an issue.

Reply:
> @Roms1383 you have to make sure that the gamedataItemStructure type is visible from C++, which means you might need to include an additional file by adding something like
include!("RED4ext/Scripting/Natives/Generated/game/data/ItemStructure_Record.hpp")

So searching for them in `RED4ext` repo doesn't yield any result,
it does in `RED4ext.SDK` but its representation is just a `uint_8`: see [over there.](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/NativeTypes.hpp#L105-L113)

I tried what you suggested but cannot make sense of the error messages on build. I probably didn't add it correctly.

Reply:
> If you want to add IsValid to ItemID I think it'd be preferable to implement it manually if it's simple enough.
When I talked about calling IsValid using the macro I just meant that a user can still pretty easily call the method if it's not included in the crate.

So for this one @jac3km4 it's quite weird because I've searched for the implementation of `IsValid` everywhere and haven't found any across `CET` / `RED4ext.SDK`. Only hint is the `native` declaration in adamsmasher's sources.

Reply:
> include!("RED4ext/Scripting/Natives/Generated/game/data/ItemStructure_Record.hpp")
I'm a bit stuck atm when trying to use `ItemStructure_Record` too.

Reply:
I can try to fix it and push to your branch

Reply:
feel free ! maybe I can learn a thing or 2 from this PR :)

Reply:
I've assumed that red4ext already had these types, but they don't seem to exist there, so we can't import them. I'd opt to keep the representation the same as red4ext for ease of maintenance, and instead provide methods that convert the `u8` fields to enums.

Reply:
After a bit of debugging, I think the ItemID creation logic is what's responsible for validating the TDBID. It doesn't seem trivial to do, so maybe it can be added later.

Reply:
I can merge the PR if you're happy with it

Reply:
> After a bit of debugging, I think the ItemID creation logic is what's responsible for validating the TDBID. It doesn't seem trivial to do, so maybe it can be added later.

You could find where it's defined ?

Reply:
> I can merge the PR if you're happy with it

If you're happy with it, I'm happy with it 😌